aiohttp/web_response.py

```diff
@@ -364,32 +364,52 @@ async def _prepare_headers(self) -> None:
         if self._compression:
             await self._start_compression(request)
 
+        # The following should not have a body per
+        # https://datatracker.ietf.org/doc/html/rfc9112#section-6.3
+        # status code: 204, 304, 1xx
+        # method: HEAD
+        request_indicates_empty_body = (
+            request.method == "HEAD"
+            or self.status in (204, 304)
+            or 100 <= self.status < 200
+        )
+
         if self._chunked:
             if version != HttpVersion11:
                 raise RuntimeError(
                     "Using chunked encoding is forbidden "
                     "for HTTP/{0.major}.{0.minor}".format(request.version)
                 )
-            writer.enable_chunking()
-            headers[hdrs.TRANSFER_ENCODING] = "chunked"
+            if not request_indicates_empty_body:
+                writer.enable_chunking()
+                headers[hdrs.TRANSFER_ENCODING] = "chunked"
+            elif hdrs.TRANSFER_ENCODING in headers:
+                # remove transfer codings when they are not needed
+                # per https://datatracker.ietf.org/doc/html/rfc9112#section-6.1
+                del headers[hdrs.TRANSFER_ENCODING]
             if hdrs.CONTENT_LENGTH in headers:
                 del headers[hdrs.CONTENT_LENGTH]
         elif self._length_check:
             writer.length = self.content_length
             if writer.length is None:
-                if version >= HttpVersion11 and self.status != 204:
-                    writer.enable_chunking()
-                    headers[hdrs.TRANSFER_ENCODING] = "chunked"
+                if version >= HttpVersion11:
+                    if not request_indicates_empty_body:
+                        writer.enable_chunking()
+                        headers[hdrs.TRANSFER_ENCODING] = "chunked"
                     if hdrs.CONTENT_LENGTH in headers:
                         del headers[hdrs.CONTENT_LENGTH]
                 else:
                     keep_alive = False
             # HTTP 1.1: https://tools.ietf.org/html/rfc7230#section-3.3.2
             # HTTP 1.0: https://tools.ietf.org/html/rfc1945#section-10.4
-            elif version >= HttpVersion11 and self.status in (100, 101, 102, 103, 204):
+            elif version >= HttpVersion11 and request_indicates_empty_body:
                 del headers[hdrs.CONTENT_LENGTH]
+                # remove transfer codings when they are not needed
+                # per https://datatracker.ietf.org/doc/html/rfc9112#section-6.1
+                if hdrs.TRANSFER_ENCODING in headers:
+                    del headers[hdrs.TRANSFER_ENCODING]
```

Comment:
Do you reckon we can just move this out a level and remove the duplicate check at lines 386-389?

```suggestion

        if version >= HttpVersion11 and request_indicates_empty_body:
            # remove transfer codings when they are not needed
            # per https://datatracker.ietf.org/doc/html/rfc9112#section-6.1
            if hdrs.CONTENT_LENGTH in headers:
                del headers[hdrs.CONTENT_LENGTH]
            if hdrs.TRANSFER_ENCODING in headers:
                del headers[hdrs.TRANSFER_ENCODING]
```

Reply:
I think so.  Once I reconcile the other issue I mentioned below I'll do some testing with the suggested change

Reply:
done in 5b557b983ed07c6ebec2c0dfbcd688b8d0ae6c8b